### PR TITLE
Switch to using bokeh_sampledata package

### LINKED ISSUE #13856: Move sampledata files to pip/conda installable package
Moving the sampledata to an  installable package will afford the following:

* `bokeh_sampledata` can depend on `pandas` potentially allowing the main package to drop `pandas`
* having an installable package may help with `pyodide` usage (however, it will be a large package... likely tens of MB)
* custom S3 download machinery can be removed from the main repo

I think the simplest thing to do will be to keep all the existing `bokeh.sampledata.foo` modules as shims, and the ones that need to can try to access their data from the separate `bokeh_sampledata` module. If `bokeh_sampledata` cannot be imported then an actionable error can be raised. This approach will avoid any changes to examples and docs, which would otherwise be significant if we want to rip `bokeh.sampledata` out of the main package entirely. 

If any of the "small" datasets currently shipped with the package depend on `pandas` they will need to be moved. Or maybe we just move all the actual data, but keep the `bokeh.sampledata` shim module structure intact. 

cc @philippjfr

### PR DESCRIPTION
**Pull request by bryevdv**
fixes: #13856 

Initial push to see how CI behaves. cc @philippjfr 

@bokeh/dev For simplicity I opted to just mention `pip install bokeh_sampledata` in most places (error messages especially) but we could expand to include `conda` if that seems necessary.

Additionally, for now I have just removed everything related to `bokeh sampledata` altogether. But can add some kind of actionable error message in case users attempt to run that, if desired.

### CODE REVIEW

File: conda/environment-test-minimal-deps.yml [hunk 1]
```diff
@@ -55,6 +55,7 @@ dependencies:
   - pip:
     - mypy >=1.10
     # docs
+    # omit bokeh_sampledata
```
Comment (bryevdv) [inline conda/environment-test-minimal-deps.yml:58]: It jsut occurs to me, why are we installing anything docs related in the "minimal" env? Perhaps all of these can be removed?

**Comment by hoxbro** [comment]
This is convenience more than anything else, but [LSP](https://en.wikipedia.org/wiki/Language_Server_Protocol), like `pylance` / `pyright`  (what a lot of VSCode user use), works differently than how `IPython` does. 

LSPs do not actually run the code, so they cannot access the values of `__dir__` like `IPython` does.  So if you want to autocomplete `from bokeh.sampledata.us_cities import d` with an LSP you need to have the actual import (guarded behind `TYPE_CHECKING`) for autocomplete to work.

Here I have an example without the import:
image

and with the import:
image

Where I have updated `us_citites.py` to:
``` python
from typing import TYPE_CHECKING
from . import _create_sampledata_shim

if TYPE_CHECKING:
    from bokeh_sampledata.us_cities import data  # noqa

__getattr__, __dir__ = _create_sampledata_shim(__name__)
```

I'm not saying this is needed, but at least something to be aware of.

**Comment by bryevdv** [comment]
@Hoxbro That makes sense, and as an aside, is coincidentally very closely aligned with the new discussion https://github.com/bokeh/bokeh/discussions/13870. Would it be possible to take a similar approach as described there, and generate `.pyi` stubs to include instead? I ask because adding all those guarded imports directly in the main repo modules will make for a very large "manually keep it in sync" surface area, whereas I suspect stub files could be auto-generated. 

(I would very much like to have any comments from you in #13870  as well)

**Comment by bryevdv** [comment]
All the examples seem to be working with the shims, but currently the docs for the shim modules is not:

Screenshot 2024-05-06 at 17 46 38

**Comment by bryevdv** [comment]
Docs are fixed now, I think this is ready for review. @tcmetzger let me know what you think about any docs needs. 

@Hoxbro assuming you agree about `.pyi` file approach, we can add that in another follow-on PR to apply that across the repo wherever it makes sense.

**Comment by hoxbro** [comment]
> generate .pyi stubs to include instead?

Yes, this should be possible. I did some initial testing below.

I have created the following folder structure to test it out:

image

Where `us_cities.pyi` contains the following:

``` python
from __future__ import annotations

data: callable

__dir__: list[str]
```

And that is autocompleting:
image

I think the logic follows [PEP-561](https://peps.python.org/pep-0561/), and I think adding the stubs folder will take precedence over the `bokeh` folder. 

Without the stubs folder:
image

With the stubs folder:
image


So then I tried to add the `us_cities.pyi` directly to `bokeh/sampledata`, which also looks to work.

File: src/bokeh/sampledata/__init__.py [hunk 2]
```diff
@@ -50,22 +20,22 @@
...
-)
+__all__ = ()
+
+SAMPLEDATA_MIN_VERSION = "2024.2"
 
 #-----------------------------------------------------------------------------
 # General API
 #-----------------------------------------------------------------------------
 
-from ..util.sampledata import download
```
Comment (hoxbro) [inline src/bokeh/sampledata/__init__.py:67]: I have used `download` in the Holoviz CI. So maybe have something like this (the language can change as needed):

``` python
def download():
    from warnings import warn
    warn("download is no longer needed, as all data is available in the `bokeh_sampledata` package", BokehUserWarning) 
```

This is mainly for people who have already installed the new package and don't understand why they can no longer download the data.

**Comment by bryevdv** [comment]
cc @mosc9575 also if you have any comments about the docs or anything else

File: src/bokeh/sampledata/__init__.py (hunk 2 above)
Comment (bryevdv) [reply src/bokeh/sampledata/__init__.py:67]: https://github.com/bokeh/bokeh/pull/13874/commits/094987a8e11b31157f7e3ba6582599bb0b8550e0

**Comment by bryevdv** [comment]
@Hoxbro that's really encouraging. I had imagined putting the `.pyi` files alongside the source files, which seems to be what the PEP suggests, if you are not also creating a stub-only distribution
> For package maintainers wishing to ship stub files containing all of their type information, it is preferred that the *.pyi stubs are alongside the corresponding *.py files.

Is there any specific advantages to using the separate `bokeh-stubs` directory? ("keeping them separate" is both a plus and a minus for organization and maintenance, but wondering if there are any other overriding considerations)

**Comment by hoxbro** [comment]
> I had imagined putting the `.pyi` files alongside the source files

Sorry, I can see I wasn't clear in my previous post. When using `.pyi` files alongside the original files. You get the behavior like without the stubs folder in the example above, e.g., you don't need to create a `.pyi` file for every `.py` file for the LSP to figure out the structure. So this is what I would do. 

> Is there any specific advantages to using the separate bokeh-stubs directory?

Not as far as I see. This was just the starting point I took to test things out.

**Review by mattpap — APPROVED** [review]
Looks good. Let's follow up with `*.pyi`, etc., in future PRs.

**Comment by bryevdv** [comment]
OK I think there's still some docs tweaks and tests that can be added but let's get this merged so that @philippjfr can experiment with a dev build

**Pull request merged**